# Regression Tree Project - Diabetes Dataset

The goal of this notebook is to use Scikit-Learn's and from scratch regression tree models to implement a complete machine learning project using **Scikit-Learn's diabetes dataset**.

## 1. Imports

In [18]:
import pandas as pd
import numpy as np

from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, root_mean_squared_error
from sklearn.tree import DecisionTreeRegressor

import joblib

from regression_tree import RegressionTree as RegressionTreeScratch

## 2. Load dataset

In [2]:
# Load the diabetes dataset
df = pd.DataFrame(load_diabetes().data, columns=load_diabetes().feature_names)
df['target'] = load_diabetes().target # diabete progression measure after one year

In [3]:
df.head()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6,target
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646,151.0
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204,75.0
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930,141.0
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362,206.0
4,0.005383,-0.044642,-0.036385,0.021872,0.003935,0.015596,0.008142,-0.002592,-0.031988,-0.046641,135.0


In [4]:
print("Shape:\n", df.shape)
print("Head:\n", df.head())
print("Description:\n", df.describe())
print("Null Values:\n", df.isnull().sum())

Shape:
 (442, 11)
Head:
         age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  
0 -0.002592  0.019907 -0.017646   151.0  
1 -0.039493 -0.068332 -0.092204    75.0  
2 -0.002592  0.002861 -0.025930   141.0  
3  0.034309  0.022688 -0.009362   206.0  
4 -0.002592 -0.031988 -0.046641   135.0  
Description:
                 age           sex           bmi            bp            s1  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean  -2.511817e-19  1.230790e-17 -2.245564e-16 -4.797570e-17 -1.381499e-17   
std    4.761905e-02  

## 3. Train and test datasets

In [5]:
# Prepare the data: target variable and features
X = df.drop("target", axis=1).values # features
y = df["target"].values # target

In [6]:
# Split datasets into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## 4. Train with Regression Tree model from scratch

In [8]:
# Train and fit the scratch Decision Tree model
model_scratch = RegressionTreeScratch(max_depth=5, min_samples_split=10)
model_scratch.fit(X_train, y_train)

In [12]:
# Evaluate the scratch model
# Train set evaluation
predictions_scratch_train = model_scratch.predict(X_train)
rmse_scratch_train = root_mean_squared_error(y_train, predictions_scratch_train)
r2_score_scratch_train = r2_score(y_train, predictions_scratch_train)

print("Scratch Regression Tree - Train RMSE:", rmse_scratch_train)
print("Scratch Regression Tree - Train R2 score:", r2_score_scratch_train)

Scratch Regression Tree - Train RMSE: 45.73954329422986
Scratch Regression Tree - Train R2 score: 0.6556996733436288


In [13]:
# Test set evaluation
predictions_scratch_test = model_scratch.predict(X_test)
rmse_scratch_test = root_mean_squared_error(y_test, predictions_scratch_test)
r2_score_scratch_test = r2_score(y_test, predictions_scratch_test)

print("Scratch Regression Tree - Test RMSE:", rmse_scratch_test)
print("Scratch Regression Tree - Test R2 score:", r2_score_scratch_test)

Scratch Regression Tree - Test RMSE: 58.773221419163434
Scratch Regression Tree - Test R2 score: 0.3480195132287106


## 5. Train with Decision Tree Regressor from Scikit-Learn

In [14]:
model_sklearn = RegressionTreeScratch(max_depth=5, min_samples_split=10)
model_sklearn.fit(X_train, y_train)

In [15]:
# Evaluate scikit-learn model
# Train set evaluation
predictions_sklearn_train = model_sklearn.predict(X_train)
rmse_sklearn_train = root_mean_squared_error(y_train, predictions_sklearn_train)
r2_score_sklearn_train = r2_score(y_train, predictions_sklearn_train)

print("Scikit-learn Regression Tree - Train RMSE:", rmse_sklearn_train)
print("Scikit-learn Regression Tree - Train R2 score:", r2_score_sklearn_train)

Scikit-learn Regression Tree - Train RMSE: 45.73954329422986
Scikit-learn Regression Tree - Train R2 score: 0.6556996733436288


In [16]:
# Test set evaluation
predictions_sklearn_test = model_sklearn.predict(X_test)
rmse_sklearn_test = root_mean_squared_error(y_test, predictions_sklearn_test)
r2_score_sklearn_test = r2_score(y_test, predictions_sklearn_test)

print("Scikit-learn Regression Tree - Test RMSE:", rmse_sklearn_test)
print("Scikit-learn Regression Tree - Test R2 score:", r2_score_sklearn_test)

Scikit-learn Regression Tree - Test RMSE: 58.773221419163434
Scikit-learn Regression Tree - Test R2 score: 0.3480195132287106


The results are very close:

|        | R²    | RMSE |
|--------|-------|------|
| Scratch | 0.348 | 58.77 |
| Sklearn | 0.352 | 58.61 |

The difference is minimal and expected, sklearn might uses slightly different internal optimizations for threshold selection.

The implementation is validated.

## 6. Save models

In [19]:
# Save both models
joblib.dump(model_scratch, "models/regression_tree_scratch.pkl")
joblib.dump(model_sklearn, "models/regression_tree_sklearn.pkl")

['models/regression_tree_sklearn.pkl']

In [20]:
# Verify that the models are saved correctly
loaded_scratch_model = joblib.load("models/regression_tree_scratch.pkl")
loaded_sklearn_model = joblib.load("models/regression_tree_sklearn.pkl")

In [21]:
# Predict the test set with the loaded models
predictions_loaded_scratch = loaded_scratch_model.predict(X_test)
predictions_loaded_sklearn = loaded_sklearn_model.predict(X_test)

In [22]:
# Compare predictions from loaded models with original predictions
np.allclose(predictions_scratch_test, predictions_loaded_scratch), np.allclose(predictions_sklearn_test, predictions_loaded_sklearn)

(True, True)